### ISSUE TITLE
No instance for (KnownSymbol ("a" ++ "b"))

### ISSUE THREAD

**Issue by mchaver** [haskell-servant/servant#864]
Is there a way to support this or extend servant such that `"a" ++ "b"` gets treated as `/ab` and `"a" ++ "b" ++ "c"` gets treated as `/abc`, etc.? The latest version of base has `AppendSymbol`, which I think should be able to avoid this because the compiler handles it, but I would like to be able to support a wider range of GHC versions.

https://github.com/ekmett/constraints/blob/master/src/Data/Constraint/Symbol.hs#L44
```haskell
import Data.Constraint.Symbol (type (++))

type X = ("a" ++ "b") :> Get '[JSON] Text

xServer :: Server X
xServer = pure "It worked"

xAPI :: Proxy X
xAPI = Proxy

xApp :: Application
xApp = serve xAPI xServer
```

```
    • No instance for (KnownSymbol ("a" ++ "b"))
        arising from a use of ‘serve’
    • In the expression: serve xAPI xServer
      In an equation for ‘xApp’: xApp = serve xAPI xServer
```

**Comment by alpmestan** [comment]
I tweaked your file a little bit. The main difference is that I explicitly use `Data.Constraint.Symbol.appendSymbol` and pattern match on the result to explicitly "construct" a `KnownSymbol` instance for `"a" ++ "b"`.

``` haskell
-- save this in e.g servant-append-symbol.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE ExplicitNamespaces #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
import Data.Constraint
import Data.Constraint.Symbol (type (++), appendSymbol)
import Data.Text
import GHC.TypeLits
import Network.Wai.Handler.Warp
import Servant

type X = ("a" ++ "b") :> Get '[JSON] Text

xServer :: Server X
xServer = pure "It worked"

xAPI :: Proxy X
xAPI = Proxy

xApp :: Application
xApp = case appendSymbol :: (KnownSymbol "a", KnownSymbol "b") :- KnownSymbol ("a" ++ "b") of
  Sub Dict -> serve xAPI xServer

main :: IO ()
main = run 8080 xApp
```

``` sh
# in one shell
$ nix-shell -p 'haskell.packages.ghc802.ghcWithPackages (ps: with ps; [ constraints servant servant-server wai warp ])' --run 'runhaskell code/servant-append-symbol.hs'

# in another shell
$ curl localhost:8080/ab
"It worked"
```

I don't know whether this can be made nicer. We can't write `instance (KnownSymbol a, KnownSymbol b) => KnownSymbol (a ++ b)` because the typeclass method is not exposed, instances are created on the spot by ghc IIRC. What you can do is:

``` haskell
xApp :: Application
xApp = case d of
  Sub Dict -> serve xAPI xServer

d = appendSymbol :: (KnownSymbol "a", KnownSymbol "b") :- KnownSymbol ("a" ++ "b")
```

but that's not much nicer. Maybe someone else will have a nicer solution? @phadej?

**Comment by phadej** [comment]
hehe. I guilty for `AppendSymbol` in GHC-8.2 :)

The solution @alpmestan is the first thing I'll do. It's troublesome. Another approach is to write new combinator:

```haskell
data Concat (a :: Symbol) (b :: Symbol)
```

and then write boilerplate instances like
```haskell
instance (HasServer api, KnownSymbol a, KnownSymbol b)
    => HasServer (Concat a b :> api)
  where
    serve ... = case appendSymbol ... of
        Sub Dict -> serve (Proxy :: Proxy (a ++ b :> api)) ...
```

TL;DR, you need to write some boilerplate, but you can decide where :)

**Comment by mchaver** [comment]
@alpmestan 
@phadej 
Thanks for the help. Those are very useful examples. I didn't realize that is how you use `appendSymbol`.
